src/cascadia/TerminalApp/CommandPalette.cpp

```diff
@@ -19,16 +20,138 @@ using namespace winrt::Microsoft::Terminal::Settings::Model;
 
 namespace winrt::TerminalApp::implementation
 {
+    // This is a view model class that extends the Command model,
+    // by managing a highlighted text that is computed by matching search filter characters to command name
+    FilteredCommand::FilteredCommand(Microsoft::Terminal::Settings::Model::Command const& command) :
+        _Command(command),
+        _Filter(L"")
+    {
+        _HighlightedName = _computeHighlightedName();
+
+        // Recompute the highlighted name if the command name changes
+        _commandChangedRevoker = _Command.PropertyChanged(winrt::auto_revoke, [weakThis{ get_weak() }](Windows::Foundation::IInspectable const& /*sender*/, Data::PropertyChangedEventArgs const& e) {
+            auto filteredCommand{ weakThis.get() };
+            if (filteredCommand && e.PropertyName() == L"Name")
+            {
+                filteredCommand->HighlightedName(filteredCommand->_computeHighlightedName());
+            }
+        });
+    }
+
+    void FilteredCommand::UpdateFilter(winrt::hstring const& filter)
+    {
+        // If the filter was not changed we want to prevent the re-computation of matching
+        // that might result in triggering a notification event
+        if (filter != _Filter)
+        {
+            Filter(filter);
+            HighlightedName(_computeHighlightedName());
+        }
+    }
+
+    // Method Description:
+    // - Looks up the filter characters within the command name.
+    // Iterating through the filter and the command name it tries to associate the next filter character
+    // with the first appearance of this character in the command name suffix.
+    //
+    // E.g., for filter="c l t s" and name="close all tabs after this", the match will be "CLose TabS after this".
+    //
+    // The command name is then split into segments (groupings of matched and non matched characters).
+    //
+    // E.g., the segments were the example above will be "CL", "ose ", "T", "ab", "S", "after this".
+    //
+    // The segments matching the filter characters are marked as highlighted.
+    //
+    // E.g., ("CL", true) ("ose ", false), ("T", true), ("ab", false), ("S", true), ("after this", false)
+    //
+    // TODO: we probably need to merge this logic with _getWeight computation?
```

Comment:
@zadjii-msft - it is actually O(m+n) - we either advance in the filter or in the name. But we definitely should not do it twice, not because of the performance, but rather because of consistency. I will try to unify the logic. If it somehow becomes complex I will open a followup task.

Reply:
@zadjii-msft - I modified the weight computation logic to use the result of segments computation and moved everything related to filtering and comparison out of CommandPalette into FilteredCommand. Hope this is aligned with @DHowett 's vision.

Actually I tried to replace everything related to sorting and filtering out of code-behind, using CollectionViewSource, to discover it is only partially supported in WinRT. So I parked there. But I believe we can create an advanced CollectionViewSource of our own or to use one from the toolkits (if this is considered appropriate) for further cleanup of the CommandPalette.